lib/cucumber/formatter/json.rb

```diff
@@ -1,19 +1,280 @@
-    class Json < GherkinFormatterAdapter
+    class Json
       include Io
 
-      def initialize(runtime, io, options)
-        @io = ensure_io(io, "json")
-        super(Gherkin::Formatter::JSONFormatter.new(@io), false, options)
+      def initialize(runtime, io, _options)
+        @runtime = runtime
+        @io = ensure_io(io, 'json')
+        @feature_hashes = []
+      end
+
+      def before_test_case(test_case)
+        builder = Builder.new(test_case)
+        unless same_feature_as_previous_test_case?(test_case.feature)
+          @feature_hash = builder.feature_hash
+          @feature_hashes << @feature_hash
+        end
+        @test_case_hash = builder.test_case_hash
+        if builder.background?
+          feature_elements << builder.background_hash
+          @element_hash = builder.background_hash
+        else
+          feature_elements << @test_case_hash
+          @element_hash = @test_case_hash
+        end
+      end
+
+      def before_test_step(test_step)
+        return if prepare_world_hook?(test_step)
+        if hook?(test_step)
+          @step_or_hook_hash = {}
+          hooks_of_type(test_step) << @step_or_hook_hash
+          return
+        end
+        if first_step_after_background?(test_step)
+          feature_elements << @test_case_hash
+          @element_hash = @test_case_hash
+        end
+        @step_or_hook_hash = create_step_hash(test_step.source.last)
+        steps << @step_or_hook_hash
+        @step_hash = @step_or_hook_hash
+      end
+
+      def after_test_step(test_step, result)
+        return if prepare_world_hook?(test_step)
+        add_match_and_result(test_step.source.last, result)
+      end
+
+      def done
+        @io.write(MultiJson.dump(@feature_hashes, pretty: true))
+      end
+
+      def puts(message)
+        test_step_output << message
+      end
+
+      def embed(src, mime_type, _label)
+        if File.file?(src)
+          content = File.open(src, 'rb') { |f| f.read }
+          data = encode64(content)
+        else
+          if mime_type =~ /;base64$/
+            mime_type = mime_type[0..-8]
+            data = src
+          else
+            data = encode64(src)
+          end
+        end
+        test_step_embeddings << { mime_type: mime_type, data: data }
+      end
+
+      private
+
+      def same_feature_as_previous_test_case?(feature)
+        current_feature[:uri] == feature.file && current_feature[:line] == feature.location.line
+      end
+
+      def first_step_after_background?(test_step)
+        test_step.source[1].name != @element_hash[:name]
+      end
+
+      def prepare_world_hook?(test_step)
+        test_step.source.last.location.file.end_with?('cucumber/filters/prepare_world.rb')
+      end
+
+      def hook?(test_step)
+        hook_source?(test_step.source.last)
+      end
+
+      def hook_source?(step_source)
+        ['Before hook', 'After hook', 'AfterStep hook'].include? step_source.name
```

Comment:
The trouble with having predicates like that is we'd really need to add them to every single AST object so that all source elements have the same protocol.

The intention is that you call `describe_source_to` passing some other object, and that object will receive `hook` or `step` messages depending on the source of the step. What is it that you need the `hook_source?` method for?